Invalid suggestion for `explicit_iter_loop`, which results in compiler error [trait bounds]

Code to reproduce the error
```rust
#![deny(clippy::explicit_iter_loop)]

fn takes_iterator<T>(iterator: &T)
where
    for<'a> &'a T: IntoIterator<Item = &'a String>
{
    for i in iterator.into_iter() {
        println!("{}", i);
    }
}


fn main() {
    let iterator = vec!["hello".to_string(), "world".to_string(), "yeah".to_string()];

    dbg!(takes_iterator(&iterator));
}
```
 [playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&code=%23!%5Bdeny(clippy%3A%3Aexplicit_iter_loop)%5D%0A%0Afn%20takes_iterator%3CT%3E(iterator%3A%20%26T)%0Awhere%0A%20%20%20%20for%3C%27a%3E%20%26%27a%20T%3A%20IntoIterator%3CItem%20%3D%20%26%27a%20String%3E%0A%7B%0A%20%20%20%20for%20i%20in%20iterator.into_iter()%20%7B%0A%20%20%20%20%20%20%20%20println!(%22%7B%7D%22%2C%20i)%3B%0A%20%20%20%20%7D%0A%7D%0A%0A%0Afn%20main()%20%7B%0A%20%20%20%20let%20iterator%20%3D%20vec!%5B%22hello%22.to_string()%2C%20%22world%22.to_string()%2C%20%22yeah%22.to_string()%5D%3B%0A%0A%20%20%20%20dbg!(takes_iterator(%26iterator))%3B%0A%7D%0A)

Output from `clippy`:
```
    Checking playground v0.0.1 (/playground)
error: it is more concise to loop over references to containers instead of using explicit iteration methods
 --> src/main.rs:7:14
  |
7 |     for i in iterator.into_iter() {
  |              ^^^^^^^^^^^^^^^^^^^^ help: to write this more concisely, try: `&iterator`
  |
note: lint level defined here
 --> src/main.rs:1:9
  |
1 | #![deny(clippy::explicit_iter_loop)]
  |         ^^^^^^^^^^^^^^^^^^^^^^^^^^
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#explicit_iter_loop

error: aborting due to previous error

error: could not compile `playground`.

To learn more, run the command again with --verbose.
```


Following this suggestion would result in a compiler error:
```rust
#![deny(clippy::explicit_iter_loop)]

fn takes_iterator<T>(iterator: &T)
where
    for<'a> &'a T: IntoIterator<Item = &'a String>
{
    for i in &iterator {
        println!("{}", i);
    }
}
```
[playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&code=%23!%5Bdeny(clippy%3A%3Aexplicit_iter_loop)%5D%0A%0Afn%20takes_iterator%3CT%3E(iterator%3A%20%26T)%0Awhere%0A%20%20%20%20for%3C%27a%3E%20%26%27a%20T%3A%20IntoIterator%3CItem%20%3D%20%26%27a%20String%3E%0A%7B%0A%20%20%20%20for%20i%20in%20%26iterator%20%7B%0A%20%20%20%20%20%20%20%20println!(%22%7B%7D%22%2C%20i)%3B%0A%20%20%20%20%7D%0A%7D%0A%0A%0Afn%20main()%20%7B%0A%20%20%20%20let%20iterator%20%3D%20vec!%5B%22hello%22.to_string()%2C%20%22world%22.to_string()%2C%20%22yeah%22.to_string()%5D%3B%0A%0A%20%20%20%20dbg!(takes_iterator(%26iterator))%3B%0A%7D%0A)
```
   Compiling playground v0.0.1 (/playground)
error[E0277]: `&&T` is not an iterator
 --> src/main.rs:7:14
  |
7 |     for i in &iterator {
  |              ^^^^^^^^^ `&&T` is not an iterator
  |
  = help: the trait `std::iter::Iterator` is not implemented for `&&T`
  = note: required by `std::iter::IntoIterator::into_iter`

error: aborting due to previous error

For more information about this error, try `rustc --explain E0277`.
error: could not compile `playground`.

To learn more, run the command again with --verbose.
```

The correct suggestion would be to replace `iterator.into_iter()` with `iterator` (which does make sense, because the for loop is syntax sugar for that)

```rust
#![deny(clippy::explicit_iter_loop)]

fn takes_iterator<T>(iterator: &T)
where
    for<'a> &'a T: IntoIterator<Item = &'a String>
{
    for i in iterator {
        println!("{}", i);
    }
}
```
[playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&code=%23!%5Bdeny(clippy%3A%3Aexplicit_iter_loop)%5D%0A%0Afn%20takes_iterator%3CT%3E(iterator%3A%20%26T)%0Awhere%0A%20%20%20%20for%3C%27a%3E%20%26%27a%20T%3A%20IntoIterator%3CItem%20%3D%20%26%27a%20String%3E%0A%7B%0A%20%20%20%20for%20i%20in%20iterator%20%7B%0A%20%20%20%20%20%20%20%20println!(%22%7B%7D%22%2C%20i)%3B%0A%20%20%20%20%7D%0A%7D%0A%0A%0Afn%20main()%20%7B%0A%20%20%20%20let%20iterator%20%3D%20vec!%5B%22hello%22.to_string()%2C%20%22world%22.to_string()%2C%20%22yeah%22.to_string()%5D%3B%0A%0A%20%20%20%20dbg!(takes_iterator(%26iterator))%3B%0A%7D%0A)

--- a/clippy_lints/src/loops.rs
+++ b/clippy_lints/src/loops.rs
@@ -12,8 +12,7 @@ use rustc_session::declare_tool_lint;
 // use rustc::middle::region::CodeExtent;
 use crate::consts::{constant, Constant};
 use crate::utils::usage::mutated_variables;
-use crate::utils::{is_type_diagnostic_item, qpath_res, sext, sugg};
-use rustc::ty::subst::Subst;
+use crate::utils::{is_type_diagnostic_item, qpath_res, same_tys, sext, sugg};
 use rustc::ty::{self, Ty};
 use rustc_data_structures::fx::{FxHashMap, FxHashSet};
 use rustc_errors::Applicability;
@@ -1344,20 +1343,9 @@ fn check_for_loop_arg(cx: &LateContext<'_, '_>, pat: &Pat<'_>, arg: &Expr<'_>, e
                     lint_iter_method(cx, args, arg, method_name);
                 }
             } else if method_name == "into_iter" && match_trait_method(cx, arg, &paths::INTO_ITERATOR) {
-                let def_id = cx.tables.type_dependent_def_id(arg.hir_id).unwrap();
-                let substs = cx.tables.node_substs(arg.hir_id);
-                let method_type = cx.tcx.type_of(def_id).subst(cx.tcx, substs);
-
-                let fn_arg_tys = method_type.fn_sig(cx.tcx).inputs();
-                assert_eq!(fn_arg_tys.skip_binder().len(), 1);
-                if fn_arg_tys.skip_binder()[0].is_region_ptr() {
-                    match cx.tables.expr_ty(&args[0]).kind {
-                        // If the length is greater than 32 no traits are implemented for array and
-                        // therefore we cannot use `&`.
-                        ty::Array(_, size) if size.eval_usize(cx.tcx, cx.param_env) > 32 => {},
-                        _ => lint_iter_method(cx, args, arg, method_name),
-                    };
-                } else {
+                let receiver_ty = cx.tables.expr_ty(&args[0]);
+                let receiver_ty_adjusted = cx.tables.expr_ty_adjusted(&args[0]);
+                if same_tys(cx, receiver_ty, receiver_ty_adjusted) {
                     let mut applicability = Applicability::MachineApplicable;
                     let object = snippet_with_applicability(cx, args[0].span, "_", &mut applicability);
                     span_lint_and_sugg(
@@ -1370,6 +1358,17 @@ fn check_for_loop_arg(cx: &LateContext<'_, '_>, pat: &Pat<'_>, arg: &Expr<'_>, e
                         object.to_string(),
                         applicability,
                     );
+                } else {
+                    let ref_receiver_ty = cx.tcx.mk_ref(
+                        cx.tcx.lifetimes.re_erased,
+                        ty::TypeAndMut {
+                            ty: receiver_ty,
+                            mutbl: Mutability::Not,
+                        },
+                    );
+                    if same_tys(cx, receiver_ty_adjusted, ref_receiver_ty) {
+                        lint_iter_method(cx, args, arg, method_name)
+                    }
                 }
             } else if method_name == "next" && match_trait_method(cx, arg, &paths::ITERATOR) {
                 span_lint(

--- a/tests/ui/for_loop_fixable.fixed
+++ b/tests/ui/for_loop_fixable.fixed
@@ -299,3 +299,40 @@ mod issue_2496 {
         unimplemented!()
     }
 }
+
+// explicit_into_iter_loop bad suggestions
+#[warn(clippy::explicit_into_iter_loop, clippy::explicit_iter_loop)]
+mod issue_4958 {
+    fn takes_iterator<T>(iterator: &T)
+    where
+        for<'a> &'a T: IntoIterator<Item = &'a String>,
+    {
+        for i in iterator {
+            println!("{}", i);
+        }
+    }
+
+    struct T;
+    impl IntoIterator for &T {
+        type Item = ();
+        type IntoIter = std::vec::IntoIter<Self::Item>;
+        fn into_iter(self) -> Self::IntoIter {
+            vec![].into_iter()
+        }
+    }
+
+    fn more_tests() {
+        let t = T;
+        let r = &t;
+        let rr = &&t;
+
+        // This case is handled by `explicit_iter_loop`. No idea why.
+        for _ in &t {}
+
+        for _ in r {}
+
+        // No suggestion for this.
+        // We'd have to suggest `for _ in *rr {}` which is less clear.
+        for _ in rr.into_iter() {}
+    }
+}

--- a/tests/ui/for_loop_fixable.rs
+++ b/tests/ui/for_loop_fixable.rs
@@ -299,3 +299,40 @@ mod issue_2496 {
         unimplemented!()
     }
 }
+
+// explicit_into_iter_loop bad suggestions
+#[warn(clippy::explicit_into_iter_loop, clippy::explicit_iter_loop)]
+mod issue_4958 {
+    fn takes_iterator<T>(iterator: &T)
+    where
+        for<'a> &'a T: IntoIterator<Item = &'a String>,
+    {
+        for i in iterator.into_iter() {
+            println!("{}", i);
+        }
+    }
+
+    struct T;
+    impl IntoIterator for &T {
+        type Item = ();
+        type IntoIter = std::vec::IntoIter<Self::Item>;
+        fn into_iter(self) -> Self::IntoIter {
+            vec![].into_iter()
+        }
+    }
+
+    fn more_tests() {
+        let t = T;
+        let r = &t;
+        let rr = &&t;
+
+        // This case is handled by `explicit_iter_loop`. No idea why.
+        for _ in t.into_iter() {}
+
+        for _ in r.into_iter() {}
+
+        // No suggestion for this.
+        // We'd have to suggest `for _ in *rr {}` which is less clear.
+        for _ in rr.into_iter() {}
+    }
+}

--- a/tests/ui/for_loop_fixable.stderr
+++ b/tests/ui/for_loop_fixable.stderr
@@ -130,5 +130,23 @@ error: it is more concise to loop over references to containers instead of using
 LL |     for _v in bs.iter() {}
    |               ^^^^^^^^^ help: to write this more concisely, try: `&bs`
 
-error: aborting due to 17 previous errors
+error: it is more concise to loop over containers instead of using explicit iteration methods`
+  --> $DIR/for_loop_fixable.rs:310:18
+   |
+LL |         for i in iterator.into_iter() {
+   |                  ^^^^^^^^^^^^^^^^^^^^ help: to write this more concisely, try: `iterator`
+
+error: it is more concise to loop over references to containers instead of using explicit iteration methods
+  --> $DIR/for_loop_fixable.rs:330:18
+   |
+LL |         for _ in t.into_iter() {}
+   |                  ^^^^^^^^^^^^^ help: to write this more concisely, try: `&t`
+
+error: it is more concise to loop over containers instead of using explicit iteration methods`
+  --> $DIR/for_loop_fixable.rs:332:18
+   |
+LL |         for _ in r.into_iter() {}
+   |                  ^^^^^^^^^^^^^ help: to write this more concisely, try: `r`
+
+error: aborting due to 20 previous errors
 

